File: java/com/joshmanisdabomb/lcc/entity/ai/ClassicMeleeAttackGoal.java
package com.joshmanisdabomb.lcc.entity.ai;

import net.minecraft.entity.CreatureEntity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.ai.goal.MeleeAttackGoal;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.util.EntityPredicates;

public class ClassicMeleeAttackGoal extends MeleeAttackGoal {

    private double targetX;
    private double targetY;
    private double targetZ;
    private double speed;
    private long targetTimer;

    public ClassicMeleeAttackGoal(CreatureEntity entity, double speed) {
        super(entity, speed, false);
        this.speed = speed;
    }

    @Override
    public boolean shouldExecute() {
        long i = this.attacker.world.getGameTime();
        if (i - this.targetTimer < 20L) {
            return false;
        } else {
            this.targetTimer = i;
            LivingEntity livingentity = this.attacker.getAttackTarget();
            return livingentity != null && livingentity.isAlive();
        }
    }

    @Override
    public boolean shouldContinueExecuting() {
        LivingEntity livingentity = this.attacker.getAttackTarget();
        if (livingentity == null) {
            return false;
        } else if (!livingentity.isAlive()) {
            return false;
        } else {
            return !(livingentity instanceof PlayerEntity) || !livingentity.isSpectator() && !((PlayerEntity)livingentity).isCreative();
        }
    }

    @Override
    public void startExecuting() {
        this.attacker.setAggroed(true);
    }

    @Override
    public void resetTask() {
        LivingEntity livingentity = this.attacker.getAttackTarget();
        if (!EntityPredicates.CAN_AI_TARGET.test(livingentity)) {
            this.attacker.setAttackTarget(null);
        }

        this.attacker.setAggroed(false);
    }

    @Override
    public void tick() {
        LivingEntity livingentity = this.attacker.getAttackTarget();
        this.attacker.getLookController().setLookPositionWithEntity(livingentity, 30.0F, 30.0F);

        this.targetX = livingentity.posX;
        this.targetY = livingentity.getBoundingBox().minY;
        this.targetZ = livingentity.posZ;
        this.attacker.getMoveHelper().setMoveTo(this.targetX, this.targetY, this.targetZ, this.speed);
        if (this.attacker.collidedHorizontally) this.attacker.getJumpController().setJumping();

        this.attackTick = Math.max(this.attackTick - 1, 0);

        double d0 = this.attacker.getDistanceSq(livingentity.posX, livingentity.getBoundingBox().minY, livingentity.posZ);
        this.checkAndPerformAttack(livingentity, d0);
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/entity/ai/ClassicMeleeAttackGoal.kt
package com.joshmanisdabomb.lcc.entity.ai

import com.joshmanisdabomb.lcc.extensions.isSurvival
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.ai.goal.MeleeAttackGoal
import net.minecraft.entity.mob.PathAwareEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.predicate.entity.EntityPredicates
import net.minecraft.util.Hand
import kotlin.math.max

class ClassicMeleeAttackGoal(entity: PathAwareEntity, private val speed: Double) : MeleeAttackGoal(entity, speed, false) {

    private var targetX = 0.0
    private var targetY = 0.0
    private var targetZ = 0.0
    private var lastUpdateTime = 0L
    private var field_24667 = 0

    override fun canStart(): Boolean {
        val i: Long = mob.world.time
        return if (i - lastUpdateTime < 20L) {
            false
        } else {
            lastUpdateTime = i
            val livingentity = mob.target
            livingentity != null && livingentity.isAlive
        }
    }

    override fun shouldContinue(): Boolean {
        val livingentity = mob.target
        return if (livingentity == null) {
            false
        } else if (!livingentity.isAlive) {
            false
        } else {
            livingentity !is PlayerEntity || livingentity.isSurvival
        }
    }

    override fun start() {
        mob.isAttacking = true
        field_24667 = 0
    }

    override fun stop() {
        mob.isAttacking = false
        if (!EntityPredicates.EXCEPT_CREATIVE_OR_SPECTATOR.test(mob.target ?: return)) {
            mob.target = null
        }
    }

    override fun tick() {
        val livingentity = mob.target ?: return
        mob.lookControl.lookAt(livingentity, 30.0f, 30.0f)

        this.targetX = livingentity.x
        this.targetY = livingentity.boundingBox.minY
        this.targetZ = livingentity.z
        mob.moveControl.moveTo(this.targetX, this.targetY, this.targetZ, this.speed)
        if (mob.horizontalCollision) mob.jumpControl.setActive()

        field_24667 = max(field_24667 - 1, 0)

        val d0 = mob.squaredDistanceTo(livingentity.x, livingentity.boundingBox.minY, livingentity.z)
        this.attack(livingentity, d0)
    }

    override fun attack(target: LivingEntity?, squaredDistance: Double) {
        val d = getSquaredMaxAttackDistance(target)
        if (squaredDistance <= d && this.field_24667 <= 0) {
            method_28346()
            mob.swingHand(Hand.MAIN_HAND)
            mob.tryAttack(target)
        }
    }

    override fun method_28346() { field_24667 = 20 }

    override fun method_28347() = field_24667 <= 0

    override fun method_28348() = field_24667

    override fun method_28349() = 20

}
